### PR TITLE
[MWPW-117914] Allow promobar use on top of Gnav

### PR DESCRIPTION
**Pull request by overmyheadandbody**
## Description
This allows authors to include a promo bar on top of the Global Navigation by defining a new `gnav-promo-source` metadata property with its value leading to a fragment which contains an `aside` block. For design purposes, it's important that authors use the `promobar` variation of the block, although we don't currently have a restriction for this from a code perspective.

Some additional notes:
- `utils.js` has been slightly adapted to add a modifier class to the header in case a Gnav promo was authored on a page. Its use can be seen in `style.css` where we adapt the nav height;
- some minimal changes have been made to the `aside` block CSS to ensure content alignment follows Word indentation and that the CTA does not wrap on multiple rows. The latter might be something useful for all `.con-button` elements, although I'm not aware of all the use cases;

Documentation will be added soon at https://milo.adobe.com/docs/authoring/global-navigation.

## Related Issue
Resolves: [MWPW-117914](https://jira.corp.adobe.com/browse/MWPW-117914)

## Testing instructions
As an FYI, it was decided that promos will not be available on mobile devices for the moment. Rendering will only occur if the viewport width is larger than `900px`. If a user then resizes their viewport below `900px`, the promo will be hidden.

One thing to look into is any CLS impact - the promo will be added on top of the navigation, thus pushing down all other page content. We assign a minimal value to the promo to apply the same principle as breadcrumbs - assign space early on to avoid content jumping around later. Tablet devices will likely experience _some_ CLS, since we can't guesstimate the actual length of the copy, but we're trying to keep the impact to a minimum.

## Screenshots:
**Dc Gnav with Promo**
DC page with Gnav Promo

## Test URLs
**Acrobat:**
- Before: https://main--dc--adobecom.hlx.page/drafts/ramuntea/page-with-gnav-promo-fragment?martech=off
- After: https://main--dc--adobecom.hlx.page/drafts/ramuntea/page-with-gnav-promo-fragment?milolibs=mwpw-117914-gnav-promo--milo--overmyheadandbody&martech=off

**BACOM:**
- Before: https://main--bacom--adobecom.hlx.page/drafts/ramuntea/page-with-gnav-promo-fragment
- After: https://main--bacom--adobecom.hlx.page/drafts/ramuntea/page-with-gnav-promo-fragment?milolibs=mwpw-117914-gnav-promo--milo--overmyheadandbody&martech=off

**CC:**
- Before: https://main--cc--adobecom.hlx.page/drafts/ramuntea/page-with-gnav-promo-fragment?martech=off
- After: https://main--cc--adobecom.hlx.page/drafts/ramuntea/page-with-gnav-promo-fragment?milolibs=mwpw-117914-gnav-promo--milo--overmyheadandbody&martech=off

**Milo:**
- Before: https://main--milo--overmyheadandbody.hlx.page/drafts/ramuntea/gnav-with-metadata-promo?martech=off
- After: https://mwpw-117914-gnav-promo--milo--overmyheadandbody.hlx.page/drafts/ramuntea/gnav-with-metadata-promo?martech=off

### CODE REVIEW

File: libs/blocks/aside/aside.css
Comment (Sartxi) [inline libs/blocks/aside/aside.css:483]: Just so you are aware, we saw issues copying these Word data attributes from the sidekick library. I am not sure if that is still a problem, but we had to go with `center` as a block variant as a result. More context [here](https://github.com/adobecom/milo/pull/627).

**Comment by rgclayton** [comment]
There is also a `promo' block. It was migrated from blog.adobe.com. This should probably be included?
https://github.com/adobecom/milo/tree/main/libs/blocks/promo

[Sample page](https://main--milo--adobecom.hlx.page/drafts/rclayton/blog-migration/sticky)

File: libs/blocks/global-navigation/global-navigation.js
Comment (mokimo) [inline libs/blocks/global-navigation/global-navigation.js:485]: Should we have a dynamic import for this, since it's not always used by everyone?

Or at least leave a comment+ticket so we can refactor this if we decide on a different path later on?

File: libs/blocks/global-navigation/global-navigation.js
Comment (mokimo) [inline libs/blocks/global-navigation/global-navigation.js:500]: We should add a catch either here, or even better around the whole block.

As our logic currently stands, if this promise doesn't resolve the page doesn't have a nav at all.

File: libs/blocks/global-navigation/global-navigation.js
Comment (mokimo) [inline libs/blocks/global-navigation/global-navigation.js:190]: We should create a Ticket to investigate a graceful failover strategy if any of those tasks fail - the nav will be gone

File: libs/styles/styles.css
Comment (mokimo) [inline libs/styles/styles.css:4]: Some consumers might rely on the fact it's 64px and get a tiny amount of CLS https://github.com/search?q=org%3Aadobecom+64px&type=code&p=1

Should we shrink the nav itself by 1px so we are at 64px with the border?

**Review by mokimo — APPROVED** [review]
A few comments & questions, but looks good 👍

**Comment by overmyheadandbody** [comment]
Hey @rgclayton! During the due diligence phase I also tried out the Promo block, but it lacked background image support, device-specific copy capabilities and has the close button which we don't need (and is also a bit annoying, since state is not maintained once a promo is closed), so we decided to go with the Aside block for now.

Looking into the future, this PR tailors to current Black Friday needs, but our stakeholders are still gathering up information and data to figure out whether Promos have a future on top of the Global Navigation anyway. Depending on this outcome, we'll decide how to expand support of Aside/Promo blocks inside the Global Navigation.

File: libs/blocks/aside/aside.css
Comment (overmyheadandbody) [reply libs/blocks/aside/aside.css:483]: Thanks for the context, Sean! I'm not sure this would be copied over using the Library, but it also doesn't hurt to have it. I'll look into adding the `center` variant for the `promobar` variant as well.

File: libs/blocks/global-navigation/global-navigation.js
Comment (overmyheadandbody) [reply libs/blocks/global-navigation/global-navigation.js:485]: We'd still need some base logic here, since we need to check the metadata property and decide whether we need to import the Aside module. So essentially we'd be making an additional call for the code on L490-L509. I'm not sure it's completely worth it, wdyt?

File: libs/blocks/global-navigation/global-navigation.js
Comment (overmyheadandbody) [reply libs/blocks/global-navigation/global-navigation.js:500]: Good idea, I'll see what makes sense to wrap so we avoid any issues 👍

File: libs/blocks/global-navigation/global-navigation.js
Comment (overmyheadandbody) [reply libs/blocks/global-navigation/global-navigation.js:190]: Well, they rely on each other anyway, so essentially if anything before `this.decorateTopnavWrapper` fails we wouldn't show a navigation. It wouldn't make sense to just show some bits and pieces that actually succeeded, like a nav with just Search and Breadcrumbs, either. IMS has its own catch, while `this.addChangeEventListeners` has all the safety checks in place. Lastly, the whole `init` method is wrapped in a `logErrorFor` call, so I think there should be enough safety mechanisms in place. If we start seeing `"Error in global navigation init"` in our logs, we can look into this more.

File: libs/blocks/global-navigation/global-navigation.js
Comment (mokimo) [reply libs/blocks/global-navigation/global-navigation.js:485]: It's a though one, if it grows certainly own module - for now we can keep it

File: libs/styles/styles.css
Comment (overmyheadandbody) [reply libs/styles/styles.css:4]: I need to look into this a bit more, good callout

File: libs/blocks/aside/aside.css
Comment (overmyheadandbody) [reply libs/blocks/aside/aside.css:483]: Adding the `center` variation skews the layout because of https://github.com/adobecom/milo/blob/main/libs/blocks/aside/aside.css#L48-L50 interacting with https://github.com/adobecom/milo/blob/main/libs/blocks/aside/aside.css#L634-L636.

I'm not aware of all the Aside use-cases, as it seems highly complex, so I think I'll stay away from additional logic for the moment. One could argue that if `:not(.center)` is used on L634 (`.aside:not(.notification):not(.inline):not(.center) .foreground.container> :last-child`), then it could also be used on L48 (`.aside:not(.notification):not(.inline) .foreground> :last-child`), which would fix the layout issues I'm seeing, but I wouldn't want to make assumptions.

File: libs/blocks/global-navigation/global-navigation.js
Comment (narcis-radu) [inline libs/blocks/global-navigation/global-navigation.js:187]: I was looking at @mokimo's suggestion to dynamically import `decorateAside` logic, maybe we can use something like
```js
const tasks = [ loadBaseStyles, this.decorateMainNav, thisDecorateTopNav, this.decorateTopNavWrapper, ... ];
if (getMetadata('gnav-promo-source')) {
  this.decorateAside = await import('.../path');
  tasks.splice(2, 0, this.decorateAside);
}
```

Comment (narcis-radu) [reply libs/blocks/global-navigation/global-navigation.js:187]: I know we would be making an additional call, but might be worth it vs. loading the code and never use it.

Comment (overmyheadandbody) [reply libs/blocks/global-navigation/global-navigation.js:185]: I kept a hybrid approach, load minimal stuff here, but move the decoration logic to a module, https://github.com/adobecom/milo/pull/1450/commits/f82e56982923d0f4e62ad972db8f646e2a45fd68

File: libs/styles/styles.css
Comment (overmyheadandbody) [reply libs/styles/styles.css]: I think this would grant a bit more time to investigate, since no one should be manually adding `64px` in their projects and instead use `var(--global-height-nav)`. However, there are too many occurrences to enforce this rule, so I'm 'shrinking' the nav by `1px` for now.

This actually addresses a production bug, where the page content overlaps the Global Navigation for `1px.`

**Review by mokimo — APPROVED** [review]

**Review by narcis-radu — APPROVED** [review]

File: libs/blocks/global-navigation/features/aside/aside.js
Comment (overmyheadandbody) [reply libs/blocks/global-navigation/features/aside/aside.js]: Added in https://github.com/adobecom/milo/pull/1450/commits/4f5823c3e3c5ddf47e2768955391c5919fbcd8c6

**Comment by overmyheadandbody** [comment]
Hi @Sartxi! Could you please have another look over this? Its merge is pending an aside block code owner approval. Thanks!

**Review by SilviuLCF — APPROVED** [review]

**Review by Sartxi — APPROVED** [review]

**Pull request merged**
